featurizers/fingerprints.py

```diff
@@ -20,216 +21,133 @@
+
+    def _get_generator(self):
+        from rdkit.Chem.rdFingerprintGenerator import GetMorganGenerator
 
-        super().__init__(n_jobs)
-        self.radius = radius
-        self.n_bits = n_bits
-        self.use_chirality = use_chirality
-        self.use_bond_types = use_bond_types
-        self.use_features = use_features
-        self.result_type = result_type
+        return GetMorganGenerator(**self.fp_generator_kwargs)
 
     def _calculate_fingerprint(
         self, X: Union[pd.DataFrame, np.ndarray]
     ) -> np.ndarray:
-        if self.result_type == "default":
-            from rdkit.Chem.rdMolDescriptors import GetMorganFingerprint
-
-            return np.array(
-                [
-                    GetMorganFingerprint(
-                        x,
-                        radius=self.radius,
-                        useChirality=self.use_chirality,
-                        useBondTypes=self.use_bond_types,
-                        useFeatures=self.use_features,
-                    )
-                    for x in X
-                ]
-            )
-        elif self.result_type == "as_bit_vect":
-            from rdkit.Chem.rdMolDescriptors import (
-                GetMorganFingerprintAsBitVect,
-            )
-
-            return np.array(
-                [
-                    GetMorganFingerprintAsBitVect(
-                        x,
-                        radius=self.radius,
-                        useChirality=self.use_chirality,
-                        useBondTypes=self.use_bond_types,
-                        useFeatures=self.use_features,
-                        nBits=self.n_bits,
-                    )
-                    for x in X
-                ]
-            )
-        else:
-            from rdkit.Chem.rdMolDescriptors import GetHashedMorganFingerprint
-
-            return np.array(
-                [
-                    GetHashedMorganFingerprint(
-                        x,
-                        radius=self.radius,
-                        useChirality=self.use_chirality,
-                        useBondTypes=self.use_bond_types,
-                        useFeatures=self.use_features,
-                        nBits=self.n_bits,
-                    )
-                    for x in X
-                ]
-            )
+        X = self._validate_input(X)
+        return self._generate_fingerprints(X)
 
 
 class MACCSKeysFingerprint(FingerprintTransformer):
     def __init__(
         self, sparse: bool = False, n_jobs: int = 1
     ):  # the sparse parameter will be unused
-        super().__init__(n_jobs)
+        super().__init__(n_jobs=n_jobs, sparse=sparse)
 
     def _calculate_fingerprint(
         self, X: Union[pd.DataFrame, np.ndarray]
     ) -> np.ndarray:
+        X = self._validate_input(X)
         from rdkit.Chem.rdMolDescriptors import GetMACCSKeysFingerprint
 
         return np.array([GetMACCSKeysFingerprint(x) for x in X])
 
+    def _get_generator(self):
+        pass
```

Comment:
If this method is sometimes not used, then it should not be required to be overridden. Currently, this breaks the Interface Segratation Principle rule of SOLID.

Reply:
Removed. As well as `@abstractmethod` decorator in base class for the function. Is this convention a good idea? Or should we complete remove the _get_generator function from the base class?

Reply:
This method in general is ok. Just some classes use it and some don't. I think this is a fair tradeoff